### ISSUE TITLE
Deref coercion not working for `Box<Trait>`

### ISSUE THREAD

**Issue by aturon** [rust-lang/rust#22194]
The following fails to compile:

```rust
trait Trait {}

fn takes_ref(t: &Trait) {}
fn takes_box(t: Box<Trait>) {
    takes_ref(&t)
}
```

with

```
<anon>:5:15: 5:17 error: the trait `Trait` is not implemented for the type `Box<Trait>` [E0277]
<anon>:5     takes_ref(&t)
                       ^~
```

despite the `Deref` implementation [here](https://github.com/rust-lang/rust/blob/master/src/liballoc/boxed.rs#L284-L288).

I believe this should work based on the [RFC](https://github.com/rust-lang/rfcs/pull/241).

**Comment by aturon** [comment]
cc @nick29581 @wycats

**Comment by cmr** [comment]
Bump!

**Comment by steveklabnik** [comment]
Triage: today, this complains about a lack of `'static`, and I'm not 100% sure how to fix it; I thought adding `+ 'static` would, but it doesn't seem to help.

**Comment by Michael-F-Bryan** [comment]
I also encountered this bug while trying to iterate over boxed trait objects and [posted] to the user forum.

[posted]: https://users.rust-lang.org/t/funny-interaction-between-deref-coersion-and-vec-box-trait/14885/7

**Comment by Spoonbender** [comment]
Triage: no change

**Comment by nielsle** [comment]
This version seems to compile
~~~ rust
trait Trait {}
fn takes_ref(t: &dyn Trait) {}
fn takes_box(t: Box<dyn Trait>) {
    takes_ref(&*t)
}
~~~
